pkg/models/datasource_cache.go

```diff
@@ -201,22 +128,145 @@ func (ds *DataSource) GetHttpTransport() (*dataSourceTransport, error) {
 	return dsTransport, nil
 }
 
-func (ds *DataSource) sigV4Middleware(next http.RoundTripper) http.RoundTripper {
-	decrypted := ds.DecryptedValues()
+type proxyTransportCache2 struct {
+	cache map[int64]cachedRoundTripper2
+	sync.Mutex
+}
+
+type cachedRoundTripper2 struct {
+	updated      time.Time
+	roundTripper http.RoundTripper
+}
+
+var ptc2 = proxyTransportCache2{
+	cache: make(map[int64]cachedRoundTripper2),
+}
+
+func (ds *DataSource) GetHttpClient2(provider httpclient.Provider) (*http.Client, error) {
+	transport, err := ds.GetHttpTransport2(provider)
+	if err != nil {
+		return nil, err
+	}
+
+	return &http.Client{
+		Timeout:   ds.getTimeout(),
+		Transport: transport,
+	}, nil
+}
+
+func (ds *DataSource) GetHttpTransport2(provider httpclient.Provider) (http.RoundTripper, error) {
+	ptc2.Lock()
+	defer ptc2.Unlock()
+
+	if t, present := ptc2.cache[ds.Id]; present && ds.Updated.Equal(t.updated) {
+		return t.roundTripper, nil
+	}
+
+	rt, err := provider.GetTransport(ds.HTTPClientOptions())
+	if err != nil {
+		return nil, err
+	}
+
+	ptc2.cache[ds.Id] = cachedRoundTripper2{
+		roundTripper: rt,
+		updated:      ds.Updated,
+	}
 
-	return sigv4.New(
-		&sigv4.Config{
+	return rt, nil
+}
+
+func (ds *DataSource) HTTPClientOptions() sdkhttpclient.Options {
+	tlsOptions := ds.TLSOptions()
+	opts := sdkhttpclient.Options{
+		Timeouts: &sdkhttpclient.TimeoutOptions{
+			Timeout:               ds.getTimeout(),
+			KeepAlive:             time.Duration(setting.DataProxyKeepAlive) * time.Second,
+			TLSHandshakeTimeout:   time.Duration(setting.DataProxyTLSHandshakeTimeout) * time.Second,
+			ExpectContinueTimeout: time.Duration(setting.DataProxyExpectContinueTimeout) * time.Second,
+			MaxIdleConns:          setting.DataProxyMaxIdleConns,
+			IdleConnTimeout:       time.Duration(setting.DataProxyIdleConnTimeout) * time.Second,
+		},
+		Headers: ds.getCustomHeaders(),
+		Labels: map[string]string{
+			"datasource_name": ds.Name,
+			"datasource_uid":  ds.Uid,
+			"kind":            "datasource",
+		},
+		TLS: &tlsOptions,
+	}
+
+	if ds.JsonData != nil {
+		opts.CustomOptions = ds.JsonData.MustMap()
+	}
+
+	if ds.BasicAuth {
+		opts.BasicAuth = &sdkhttpclient.BasicAuthOptions{
+			User:     ds.BasicAuthUser,
+			Password: ds.DecryptedBasicAuthPassword(),
+		}
+	} else if !ds.BasicAuth && ds.User != "" {
```

Comment:
The `if !ds.BasicAuth` here is superfluous given the condition on line 202:
```suggestion
	} else if ds.User != "" {
```